stackrox/stackrox
# ROX-20280: Unable to edit the unauthenticated SMTP server endpoint

## Description

Email notifiers have a setting `allow_unauthenticated_smtp` that allows a user to use an SMTP server without providing a password. The notifier proto has `scrub:dependent` set on the server endpoint field which prevents anyone from changing that field without providing a password (which doesn't exist). We don't want to remove that because it's a valid security check, just need to relax it for anyone who is using unauthenticated SMTP.

This change introduces a new tag to the scrubber called `scrub:disableDependentIfTrue`. If this is set on a boolean field, and the field is true then the `scrub:dependent` check is skipped. The `scrub:always` check continues to behave as-is because the credentials are supposed to be empty.

This change also adds the `scrub:disableDependentIfTrue` tag to the `allow_unauthenticated_smtp` field.

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

### Here I tell how I validated my change

Unit tests
Manual tests:
1. Deployed 4.3.4. Added unauthenticated SMTP with endpoint `test.localhost`. 
2. Tried to change it to `test.localhost:25` and it failed.
3. Upgraded to the image from this build and retried step 2. Worked.
4. Created a notifier _with_ a password.
5. Tried to edit endpoint without password. Failed as expected.
6. Tried again with password and it worked.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.
